### DISCUSSION TITLE
Best practices for SDIO devices

### DISCUSSION THREAD

**Discussion by jerome-pouiller** [fwupd/fwupd#6112] [Q&A]
Hi folks,

I plan to write a driver for an SDIO WiFi chip. This chip includes a flash containing the firmware. I would like to provide a way to update the flash.

However, I am not sure of the interface I should provide for that purpose.
It is not possible to access to the SDIO bus from the userspace. So I can't just detach the normal driver.
I have no char device in /dev. So I can't use that to send the firmware payload.

I believe I could provide auxiliary files in `/sys` to perform the upgrade. Some plugins use files in `/sys`, but I don't find any plugin using `/sys` to send the firmware payload. Do you think it is the right thing to do, or I missed something?

Thank you,

**Comment by hughsie** [comment]
Hey Jérôme! The fwupd daemon doesn't really care -- a sysfs file is just fine. What I would say is that if the firmware flash takes more and a second you probably want some way of returning status back to the user. e.g. if we can write() 1024x10Kb blocks then we can construct a progress bar that makes sense rather than just blocking for 30 seconds with one huge write().

I guess some questions:

 * Does the device need to be detached/attached to/from a bootloader mode before it'll accept a flashed image?
 * Do we need to restart the device manually when done before it'll use the new firmware?
 * How do we get the current firmware version number out of the device?
 * Does the device survive if I flash the wrong flash image to the file? e.g. if I use /dev/random does it explode?
 * How do we match the firmware stream to the device? e.g. is firmware OEM specific?

Maybe that helps?

Richard.

**Comment by jerome-pouiller** [comment]
Hi Richard,

Thank you for your answer.

> * Does the device need to be detached/attached to/from a bootloader mode before it'll accept a flashed image?

Yes, it does.

> * Do we need to restart the device manually when done before it'll use the new firmware?

Yes, the reset command must be issued by the SDIO bus controller.

> * How do we get the current firmware version number out of the device?

The driver can get it during chip initialisation and expose it in sysfs.

> * Does the device survive if I flash the wrong flash image to the file? e.g. if I use /dev/random does it explode?

I have to check that. The bootloader check at least the CRC.

> * How do we match the firmware stream to the device? e.g. is firmware OEM specific?

Really good question. We have a VID/PID and I think the firmware should work on all the devices derived from our chip. However, you know how is the business. I prefer to be prepared to have custom firmwares in the wild. I am going to check how to deal with that.

**Comment by hughsie** [comment]
>> Does the device need to be detached/attached to/from a bootloader mode before it'll accept a flashed image?
> Yes, it does.

Does it dis-attach and come back with a different VID/PID, or if not -- can we query the device to find out what "mode" it's in?
 
>> Do we need to restart the device manually when done before it'll use the new firmware?

>> Does the device survive if I flash the wrong flash image to the file? e.g. if I use /dev/random does it explode?
> I have to check that. The bootloader check at least the CRC.

Also, fwupd needs to know if the firmware blob is "signed" i.e. PKCS#7 -- it doesn't matter but it's something that bobbles all the way up to the UI in some places.
 
>> How do we match the firmware stream to the device? e.g. is firmware OEM specific?
> Really good question. We have a VID/PID and I think the firmware should work on all the devices derived from our chip. However, you know how is the business. I prefer to be prepared to have custom firmwares in the wild. I am going to check how to deal with that.

What some vendors have is a "customer ID" that we can use to build the instance ID we use for matching. e.g. `HIDRAW\VEN_17EF&DEV_7226&CID_01&BANK_1` -- if we can query that we can make sure we don't do the wrong thing.
